accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+   ;; without fields here then they will be matched on and constructed
+   ;; in field-less form (e.g. `Opcode::Iadd => { ... }`).
+   (type Opcode extern (enum Iadd Imul Isub Iconst ...))
+   
+   ;; Primitive type. These are presumably included in a prelude
+   ;; and should not normally need to be declared by a compiler
+   ;; backend author using the DSL.
+   (type u32 primitive)
+```
+
+Once we have the value types defined, we define *type signatures for
+terms*. Every term that is used -- as an internal node, as an
+extractor or as a constructor -- needs types for its arguments, and
+the whole expression is given a type (analogous to a return-value
+type).
+
+```plain
+
+    ;; `Iadd` is a term symbol, and `(Iadd ...)` terms take two arguments
+    ;; of type `Value` and produce a result of type `Value`.
+    (decl (Iadd Value Value) Value)
+    
+    (decl (Iconst u32) Value)
+```
+
+Now, given a proper type environment for values and terms, we need to
+define ways to process them: with extractors and constructors, and
+with internal rules. A particular term can *either* have an external
+extractor and/or constructor function, or can have an internal rule,
+but never both. If a term is to be defined by an extractor and/or
+constructor, it must be declared with an `extern` like so:
+
+```plain
+    (decl extern (MyTerm u32) u32)
+```
+
+## Defining External Extractors and Constructors
+
+First, we define *extractors* and *constructors* for some terms. These
+terms are those that represent the external interface to the
+instruction selector rules. Extractors, as introduced above, provide a
+view onto a virtual term tree at the input.
+
+One can think of an extractor like a "reverse function": it is given
+the value of the whole term tree that the extractor is meant to match,
+and if it matches successfully, it returns values for each of the
+arguments. Extractors are attached to terms as follows:
+
+```plain
+
+    (decl extern (concat u32 u32) u64)
+    (extractor concat "unconcat")
+```
+
+This indicates to the DSL compiler that the `concat` term can be used
+in the left-hand side of a rule (as described below), and when a match
+is evaluated, a Rust function `unconcat` will be called. This external
+function is assumed to have the signature:
+
+```rust
+
+    fn unconcat<C>(context: &mut C, value: u64) -> Option<(u32, u32)>;
+```
+
+If the function returns a `Some((a, b))` value, the match was a
+success and the arguments' values are recursively matched by
+sub-patterns.
+
+Similarly, a constructor can be attached to a term as follows:
+
+```plain
+
+    (decl extern (concat u32 u32) u64)
+    (constructor concat "concat")
+```
+
+which implies the existence of a Rust function:
+
+```rust
+
+    fn concat<C>(context: &mut C, val1: u32, val2: u32) -> u64;
+```
+
+Note that a constructor, unlike an extractor, cannot fail. More on the
+execution model below.
+
+## Defining Rules
```

Comment:
A similar, but somewhat different case here would be [`popcnt`](https://github.com/bytecodealliance/wasmtime/blob/74c18a6af4da63fa9d748e1d209b2214acbe79be/cranelift/codegen/src/isa/aarch64/lower_inst.rs#L1233), where we emit a number of instructions, and then match a single instruction depending on the type of the input.

But here, since the total size is about 6 instructions in 5 different input types it isn't as bad.

Reply:
Re-reading the RFC now, it looks like the shl case would be a good use of [Virtual Internal Terms](https://github.com/cfallin/rfcs/blob/cranelift-isel-isle-peepmatic/accepted/cranelift-isel-isle-peepmatic.md#virtual-internal-terms). `popcnt` might be too small for this to make sense.

Reply:
Yes, exactly, that's the intent of the ability to freely (in terms of runtime cost) create these virtual terms that chain rules together -- they allow additional factoring and abstraction.